xla/service/collective_quantizer.cc

```diff
@@ -76,49 +193,132 @@ bool IsSupportedQuantization(HloInstruction* instr, HloInstruction** convert,
       break;
     }
     VLOG(5) << "Unsupported instruction.";
-    return false;
+    return std::nullopt;
   }
 
   // In the quantization case, the type conversion is preceded by a
   // multiplication or division by a broadcasted scalar and a clamp instruction.
   if (ops.size() > 2 &&
       (Match(ops.back(),
-             m::Convert(convert, m::Clamp(clamp, ScalarBroadcast(),
-                                          m::MultiplyAnyOrder(
-                                              binary, m::Op(),
-                                              ScalarBroadcast(scale_bcast)),
-                                          ScalarBroadcast()))) ||
-       Match(
-           ops.back(),
-           m::Convert(convert, m::Clamp(clamp, ScalarBroadcast(),
-                                        m::Divide(binary, m::Op(),
-                                                  ScalarBroadcast(scale_bcast)),
-                                        ScalarBroadcast()))))) {
-    unary_ops = {ops.begin(), ops.end() - 3};
-  } else if (ops.size() > 0 && Match(ops.back(), m::Convert(convert))) {
-    unary_ops = {ops.begin(), ops.end() - 1};
+             m::Convert(&subgraph.convert,
+                        m::Clamp(&subgraph.clamp, ScalarBroadcast(),
+                                 m::MultiplyAnyOrder(
+                                     &subgraph.binary, m::Op(),
+                                     ScalarBroadcast(&subgraph.scale_bcast)),
```

Comment:
Moving the quantization before the collective op results in incorrect behavior if the scale is not the same on each GPU. And same if the lower and higher clamp values are not the same on each GPU.

And similarly for the dequantization case.